File: player_code/cpp/player_code.cpp
#include "player_code.h"

#include <algorithm>
#include <cmath>
#include <compare>
#include <iostream>
#include <set>
#include <sstream>
#include <unordered_map>
#include <vector>

Attributes::Attributes(unsigned hp, unsigned range, unsigned attack_power,
                       unsigned speed, unsigned price)
    : hp(hp), range(range), attack_power(attack_power), speed(speed),
      price(price) {}

Position::Position(int x, int y) : _x(x), _y(y) {}

int Position::get_x() const { return _x; }
int Position::get_y() const { return _y; }

double Position::distance_to(Position other) const {
  auto delta_x = other.get_x() - this->get_x();
  auto delta_y = other.get_y() - this->get_y();
  return sqrt((double)(delta_x * delta_x + delta_y * delta_y));
}

bool is_valid_spawn_position(int x, int y) {
  if (x < 0 || y < 0 || x >= static_cast<int>(Constants::MAP_NO_OF_COLS) ||
      y >= static_cast<int>(Constants::MAP_NO_OF_ROWS)) {
    return false;
  }
  return x == 0 || y == 0 ||
         (x == (static_cast<int>(Constants::MAP_NO_OF_COLS) - 1)) ||
         (y == (static_cast<int>(Constants::MAP_NO_OF_ROWS) - 1));
}

bool is_valid_spawn_position(Position pos) {
  return is_valid_spawn_position(pos.get_x(), pos.get_y());
}

std::vector<Position> get_all_valid_spawn_positions() {
  std::vector<Position> all_valid_positions;
  // x is 0
  for (int j = 0; j < static_cast<int>(Constants::MAP_NO_OF_ROWS); j++) {
    all_valid_positions.push_back({0, j});
  }

  // y is 0
  for (int i = 1; i < static_cast<int>(Constants::MAP_NO_OF_COLS); i++) {
    all_valid_positions.push_back({i, 0});
  }

  // x is MAP_NO_OF_ROWS-1
  for (int j = 1; j < static_cast<int>(Constants::MAP_NO_OF_ROWS); j++) {
    all_valid_positions.push_back(
        {static_cast<int>(Constants::MAP_NO_OF_COLS) - 1, j});
  }

  // y is MAP_NO_OF_COLS - 1
  for (int i = 1; (i + 1) < static_cast<int>(Constants::MAP_NO_OF_COLS); i++) {
    all_valid_positions.push_back(
        {i, static_cast<int>(Constants::MAP_NO_OF_ROWS) - 1});
  }
  return all_valid_positions;
}

Actor::Actor(size_t id, size_t hp, size_t type, Position pos)
    : _id(id), _hp(hp), _type(type), _position(pos) {}
size_t Actor::get_id() const { return _id; }
size_t Actor::get_hp() const { return _hp; }
size_t Actor::get_type() const { return _type; }
Position Actor::get_position() const { return _position; }

Attacker::Attacker(size_t id, size_t hp, size_t type, Position pos)
    : Actor(id, hp, type, pos) {}

Defender::Defender(size_t id, size_t hp, size_t type, Position pos)
    : Actor(id, hp, type, pos) {}

State::State(std::vector<Attacker> attackers, std::vector<Defender> defenders,
             size_t no_of_coins_left, size_t turn_no)
    : _turn_no(turn_no), _no_of_coins_left(no_of_coins_left),
      _attackers(std::move(attackers)), _defenders(std::move(defenders)) {}

const std::vector<Attacker> &State::get_attackers() const {
  return this->_attackers;
}
const std::vector<Defender> &State::get_defenders() const {
  return this->_defenders;
}
size_t State::get_turn_no() const { return this->_turn_no; }
size_t State::get_coins_left() const { return this->_no_of_coins_left; }

Game::Game() {}

void Game::spawn_attacker(size_t id, Position pos) {
  this->_spawn_postions.push_back({id, pos});
  this->_already_spawned_positions.insert(pos);
}
bool Game::already_spawned_at_position(Position pos) {
  return this->_already_spawned_positions.contains(pos);
}
void Game::set_target(size_t attacker_id, size_t defender_id) {
  this->_player_set_targets.insert({attacker_id, defender_id});
}
void Game::set_target(const Attacker &attacker, const Defender &defender) {
  this->_player_set_targets.insert({attacker.get_id(), defender.get_id()});
}

std::ostringstream &Game::logr() { return this->_logr; }

const std::unordered_map<size_t, size_t> &Game::get_player_set_targets() const {
  return this->_player_set_targets;
}
const std::vector<std::pair<size_t, Position>> &
Game::get_spawn_positions() const {
  return this->_spawn_postions;
}
const std::set<Position> &Game::get_already_spawned_positions() const {
  return this->_already_spawned_positions;
}

Map::Map(std::vector<std::vector<int>> map_as_grid)
    : _grid(move(map_as_grid)) {}

Map Map::get(std::istream &stream) {
  static int no_of_times_called = 0;
  no_of_times_called++;
  if (no_of_times_called > 1) {
    throw std::runtime_error(
        "Player tried to call an internal function. Not allowed");
  }

  size_t m = 0;
  size_t n = 0;
  stream >> m;
  stream >> n;

  Map::no_of_rows = m;
  Map::no_of_cols = n;

  std::vector<std::vector<int>> grid(m, std::vector<int>(n, 0));
  for (auto &row : grid) {
    for (auto &cell : row) {
      stream >> cell;
    }
  }
  return {move(grid)};
}

std::vector<Defender> Map::spawn_defenders() const {
  static int no_of_times_called = 0;
  no_of_times_called++;
  if (no_of_times_called > 1) {
    throw std::runtime_error(
        "Player tried to call an internal function. Not allowed");
  }

  std::vector<Defender> defenders;
  size_t id = 0;
  for (size_t x = 0; x < this->no_of_cols; x++) {
    for (size_t y = 0; y < this->no_of_rows; y++) {
      if (this->_grid[y][x] > 0) {
        auto &attributes_for_defender =
            Constants::DEFENDER_TYPE_ATTRIBUTES.at(this->_grid[y][x]);
        defenders.emplace_back(id++, attributes_for_defender.hp,
                               this->_grid[y][x], Position(x, y));
      }
    }
  }
  return defenders;
}

File: player_code/python/player_code.py
from dataclasses import dataclass
import sys


@dataclass(eq=True, frozen=True, order=True)
class Position:
    x: float
    y: float

    def distance_to(self, other: "Position") -> float:
        return ((self.x - other.x) ** 2 + (self.y - other.y) ** 2) ** 0.5


@dataclass(frozen=True)
class ActorType:
    hp: int
    range: int
    attack_power: int
    price: int


@dataclass(frozen=True)
class AttackerType(ActorType):
    speed: int


@dataclass(frozen=True)
class DefenderType(ActorType):
    pass


@dataclass(frozen=True)
class Attacker:
    id: int
    hp: int
    type: AttackerType
    position: Position


@dataclass(frozen=True)
class Defender:
    id: int
    hp: int
    type: DefenderType
    position: Position


@dataclass(frozen=True)
class State:
    attackers: list[Attacker]
    defenders: list[Defender]
    no_of_coins_left: int
    turn_no: int


class Game:
    def __init__(self):
        self._log = ""
        self.player_set_targets: dict[int, int] = {}
        self.spawn_positions: list[tuple[int, Position]] = []
        self.already_spawned_positions: set[Position] = set()

    def spawn_attacker(self, id: int, position: Position):
        self.spawn_positions.append((id, position))
        self.already_spawned_positions.add(position)

    def is_already_spawned_at_position(self, position: Position):
        return position in self.already_spawned_positions

    def set_target(self, attacker_id: int, defender_id: int):
        self.player_set_targets[attacker_id] = defender_id

    def log(self, line: str):
        self._log += line + "\n"

    def get_log(self):
        return self._log

    def clear_log(self):
        self._log = ""


class Constants:
    NO_OF_TURNS: int
    MAX_NO_OF_COINS: int
    NO_OF_ATTACKER_TYPES: int
    NO_OF_DEFENDER_TYPES: int
    ATTACKER_TYPE_ATTRIBUTES: dict[int, AttackerType]
    DEFENDER_TYPE_ATTRIBUTES: dict[int, DefenderType]
    MAP_NO_OF_ROWS: int
    MAP_NO_OF_COLS: int

    @classmethod
    def initialize(cls):
        cls.NO_OF_TURNS, cls.MAX_NO_OF_COINS = map(int, input().split())
        cls.NO_OF_ATTACKER_TYPES = int(input())
        cls.ATTACKER_TYPE_ATTRIBUTES = {}
        for i in range(1, cls.NO_OF_ATTACKER_TYPES + 1):
            hp, a_range, attack_power, speed, price = map(int, input().split())
            cls.ATTACKER_TYPE_ATTRIBUTES[i] = AttackerType(
                hp, a_range, attack_power, price, speed
            )

        cls.NO_OF_DEFENDER_TYPES = int(input())
        cls.DEFENDER_TYPE_ATTRIBUTES = {}
        for i in range(1, cls.NO_OF_DEFENDER_TYPES + 1):
            hp, d_range, attack_power, _, price = map(int, input().split())
            cls.DEFENDER_TYPE_ATTRIBUTES[i] = DefenderType(
                hp, d_range, attack_power, price
            )


class Map:
    map: list[list[int]]

    @classmethod
    def initialize(cls):
        Constants.MAP_NO_OF_ROWS, Constants.MAP_NO_OF_COLS = map(int, sys.stdin.readline().split())
        cls.map = []
        for _ in range(Constants.MAP_NO_OF_ROWS):
            cls.map.append(list(map(int, sys.stdin.readline().split())))

    @staticmethod
    def spawn_defenders():
        defenders = []
        id = 0
        for x in range(Constants.MAP_NO_OF_COLS):
            for y in range(Constants.MAP_NO_OF_ROWS):
                if Map.map[y][x] > 0:
                    attributes = Constants.DEFENDER_TYPE_ATTRIBUTES[Map.map[y][x]]
                    defenders.append(
                        Defender(id, attributes.hp, attributes, Position(x, y))
                    )
                    id += 1
        return defenders


def is_valid_spawn_position(pos: Position) -> bool:
    x, y = pos.x, pos.y
    if x < 0 or y < 0 or x >= Constants.MAP_NO_OF_ROWS or y >= Constants.MAP_NO_OF_COLS:
        return False
    return (
        x == 0
        or y == 0
        or x == Constants.MAP_NO_OF_ROWS - 1
        or y == Constants.MAP_NO_OF_COLS - 1
    )


def get_all_valid_spawn_positions() -> list[Position]:
    positions: set[Position] = set()
    for y in range(Constants.MAP_NO_OF_ROWS):
        positions.add(Position(0, y))
        positions.add(Position(Constants.MAP_NO_OF_COLS - 1, y))
    for x in range(Constants.MAP_NO_OF_COLS):
        positions.add(Position(x, 0))
        positions.add(Position(x, Constants.MAP_NO_OF_ROWS - 1))
    return list(positions)
